### PR TITLE
Implementing vmad shader instruction

### PR DESCRIPTION
**Pull request by Shrek5InTheatres2019**
I'm not sure if this is a proper implementation, and i know it's not complete, i couldn't figure out some of the semantics, but it's more than we had before.

### CODE REVIEW

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 1]
```diff
@@ -1,18 +1,43 @@
...
+            bool resultSigned = isSigned(op.RaType) || isSigned(op.RbType);
+
+            Operand result;
+
+            result = context.IMultiply(GetSrcA(context), GetSrcB(context));
+            result = context.IAdd(result, GetSrcC(context));
+            if (op.RawOpCode.Extract(51, 2) == 1)
+            {
+                result = resultSigned ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 7)) : context.ShiftRightU32(result, new Operand(OperandType.Constant, 7));
+            }else if (op.RawOpCode.Extract(51, 2) == 2)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:23]:
```suggestion
            }
            else if (op.RawOpCode.Extract(51, 2) == 2)
```

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 2]
```diff
@@ -1,18 +1,43 @@
...
+            {
+                result = resultSigned ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 15)) : context.ShiftRightU32(result, new Operand(OperandType.Constant, 15));
+
+            }
+            context.Copy(GetDest(context), result);
+        }
+
+        public static bool isSigned(VideoType t)
+        {
+            if(t == VideoType.S16 || t == VideoType.S8 || t == VideoType.S32)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:33]: That could be just:
```cs
return t == VideoType.S8 || t == VideoType.S16 || t == VideoType.S32;
```

File: Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs [hunk 3]
```diff
@@ -1,7 +1,12 @@
+using Ryujinx.Common.Logging;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs:1]: Why was this file changed?

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 4]
```diff
@@ -1,18 +1,43 @@
...
+            Operand result;
+
+            result = context.IMultiply(GetSrcA(context), GetSrcB(context));
+            result = context.IAdd(result, GetSrcC(context));
+            if (op.RawOpCode.Extract(51, 2) == 1)
+            {
+                result = resultSigned ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 7)) : context.ShiftRightU32(result, new Operand(OperandType.Constant, 7));
+            }else if (op.RawOpCode.Extract(51, 2) == 2)
+            {
+                result = resultSigned ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 15)) : context.ShiftRightU32(result, new Operand(OperandType.Constant, 15));
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:25]: Readability would be improved by puting the `?` and `:` on new lines.

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 5]
```diff
@@ -1,18 +1,43 @@
...
-            context.Copy(GetDest(context), GetSrcC(context));
+            OpCodeVideo op = (OpCodeVideo)context.CurrOp;
+
+            bool resultSigned = isSigned(op.RaType) || isSigned(op.RbType);
+
+            Operand result;
+
+            result = context.IMultiply(GetSrcA(context), GetSrcB(context));
+            result = context.IAdd(result, GetSrcC(context));
+            if (op.RawOpCode.Extract(51, 2) == 1)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:20]: Imo `op.RawOpCode.Extract(51, 2)` should be assigned to a variable and you should use that instead.

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 6]
```diff
@@ -1,18 +1,43 @@
...
+            {
+                result = resultSigned ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 7)) : context.ShiftRightU32(result, new Operand(OperandType.Constant, 7));
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:22]: Readability would be improved by puting the `?` and `:` on new lines.

File: Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs (hunk 3 above)
Comment (Shrek5InTheatres2019) [reply Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs:1]: When i made the branch for these changes i accidentally duplicated the master branch from my fork, which had all the ISslContext changes i had made on a different PR, so i reverted them here, but apparently not fully.

**Comment by Shrek5InTheatres2019** [comment]
Requested changes are complete

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 7]
```diff
@@ -1,18 +1,45 @@
...
+
+            int shr = op.RawOpCode.Extract(51, 2);
+
+            result = context.IMultiply(GetSrcA(context), GetSrcB(context));
+            result = context.IAdd(result, GetSrcC(context));
+
+            if (shr == 1)
+            {
+                result = resultSigned 
+                    ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 7))
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:26]: Instead of `new Operand(OperandType.Constant, 7)` you can just use `Const(7)` here. Same for the other constants.

File: Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs [hunk 8]
```diff
@@ -1,18 +1,45 @@
...
+                result = resultSigned 
+                    ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 7)) 
+                    : context.ShiftRightU32(result, new Operand(OperandType.Constant, 7));
+            }
+            else if (shr == 2)
+            {
+                result = resultSigned 
+                    ? context.ShiftRightS32(result, new Operand(OperandType.Constant, 15)) 
+                    : context.ShiftRightU32(result, new Operand(OperandType.Constant, 15));
+
```
Comment (gdkchan) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitVideo.cs:34]: New line not required.

File: Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs [hunk 9]
```diff
@@ -1,7 +1,11 @@
+using System;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslContext.cs:1]: Same comment as before.
